packages/imt-rs/src/lib.rs

```diff
@@ -0,0 +1,320 @@
+use std::error::Error;
+
+use tiny_keccak::{Hasher, Keccak};
+
+pub struct IMT {
+    nodes: Vec<Vec<Option<IMTNode>>>,
+    zeroes: Vec<IMTNode>,
+    hash: IMTHashFunction,
+    depth: usize,
+    arity: usize,
+}
+
+struct IMTMerkleProof {
+    root: IMTNode,
+    leaf: IMTNode,
+    path_indices: Vec<usize>,
+    siblings: Vec<Vec<IMTNode>>,
+    leaf_index: usize,
+}
+
+
+impl IMT {
+    pub fn new(hash: IMTHashFunction, depth: usize, zero_value: IMTNode, arity: usize, leaves: Option<Vec<IMTNode>>) -> Result<IMT, &'static str> {
+
+        if leaves.is_some() && leaves.as_ref().unwrap().len() > arity.pow(depth as u32) {
+            return Err("The tree cannot contain more than arity^depth leaves");
+        }
```

Comment:
you can get advantage of the proper [pattern matching](https://rust-book.cs.brown.edu/ch18-03-pattern-syntax.html) here
```suggestion
        if let Some(ref leaves) = leaves {
            if leaves.len() > arity.pow(depth as u32) {
                return Err("The tree cannot contain more than arity^depth leaves");
            }
        }
```